ggml-metal.metal

```diff
@@ -400,20 +436,15 @@ kernel void kernel_mul_mat_q4_0_f32(
             y_curr[i] = *((device float4  *)(y + N_SIMDWIDTH * (tiisg + column * QK4_0) + 4 * i));
             sumy += y_curr[i][0] + y_curr[i][1] + y_curr[i][2] + y_curr[i][3];
         }
-        sumy *= (-8.f);
+        // we don't right shift packed 4-bit weights, so we have to devide y by 16/256/4096 to conpensate this.
+        // this design is q4_0 and q4_1 centered, but I think most of the people use these two quantizations.
+        for (int i = 0; i < 32; i++) {
+            yl[i] *= pow(1.f/16.f, 2 * (i % 2) + i / 16);
+        }
 
         for (int row = 0; row < N_DST; row++) {
-            // prefetch next x block
-            qb_next = x[tiisg + ((row + 1) % N_DST) * nb + (column + ((row + 1) / N_DST)) * N_SIMDWIDTH];
-
-            // calculate
-            float d = qb_curr.d;
-            float acc = sumy;
-            for (int i = 0; i < 16; i++) {
-                acc += yl[i] * (qb_curr.qs[i] & 0xF) + yl[i+16] * (qb_curr.qs[i] >> 4);
-            }
-            sumf[row] += d * acc;
-            qb_curr = qb_next;
+            sumf[row] += block_q_n_dot_y(qb_curr, sumy, yl);
+            qb_curr = x[tiisg + ((row + 1) % N_DST) * nb + (column + ((row + 1) / N_DST)) * N_SIMDWIDTH];
```

Comment:
Do we gain something by having `qb_curr` set above as `x[tiisg]` outside the loop and then setting it for the next row here? (instead of just setting once inside the loop for the current row)

Reply:
Yes. For some reasons I currently don't understand. Setting `qb_curr` in the loop for the current row leads to a performance degradation from 76.5 ms/tok to 79ms/tok on 33B model.